src/ast/ast.c

```diff
@@ -0,0 +1,328 @@
+    // const char *alias = cypher_ast_identifier_get_name(cypher_ast_unwind_get_alias(unwind_clause));
+
+    // const cypher_astnode_t *ast_exp = cypher_ast_unwind_get_expression(unwind_clause);
+    // uint id = AST_AddRecordEntry(ast);
+    // AR_ExpNode *exp = AR_EXP_NewReferenceNode(alias, id, false);
+    // ast->defined_entities = array_append(ast->defined_entities, exp);
+    // ASTMap_FindOrAddAlias(ast, alias, id);
+
+// }
+
+bool AST_ReadOnly(const cypher_astnode_t *root) {
+    // Iterate over children rather than clauses, as the root is not
+    // guaranteed to be a query.
+    uint num_children = cypher_astnode_nchildren(root);
+    for (unsigned int i = 0; i < num_children; i ++) {
+        const cypher_astnode_t *child = cypher_astnode_get_child(root, i);
+        cypher_astnode_type_t type = cypher_astnode_type(child);
+        if(type == CYPHER_AST_CREATE ||
+           type == CYPHER_AST_MERGE ||
+           type == CYPHER_AST_DELETE ||
+           type == CYPHER_AST_SET ||
+           type == CYPHER_AST_CREATE_NODE_PROP_INDEX ||
+           type == CYPHER_AST_DROP_NODE_PROP_INDEX) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+// TODO is this necessary?
+bool AST_Empty(const AST *ast) {
+    return false;
+    // return cypher_ast_query_nclauses(ast->root) > 0;
+}
+
+bool AST_ContainsClause(const AST *ast, cypher_astnode_type_t clause) {
+    return AST_GetClause(ast, clause) != NULL;
+}
+
+bool AST_ContainsErrors(const cypher_parse_result_t *result) {
+    return cypher_parse_result_nerrors(result) > 0;
+}
+
+char* AST_ReportErrors(const cypher_parse_result_t *result) {
+    char *errorMsg;
+    unsigned int nerrors = cypher_parse_result_nerrors(result);
+    for(unsigned int i = 0; i < nerrors; i++) {
+        const cypher_parse_error_t *error = cypher_parse_result_get_error(result, i);
+
+        // Get the position of an error.
+        struct cypher_input_position errPos = cypher_parse_error_position(error);
+
+        // Get the error message of an error.
+        const char *errMsg = cypher_parse_error_message(error);
+
+        // Get the error context of an error.
+        // This returns a pointer to a null-terminated string, which contains a
+        // section of the input around where the error occurred, that is limited
+        // in length and suitable for presentation to a user.
+        const char *errCtx = cypher_parse_error_context(error);
+
+        // Get the offset into the context of an error.
+        // Identifies the point of the error within the context string, allowing
+        // this to be reported to the user, typically with an arrow pointing to the
+        // invalid character.
+        size_t errCtxOffset = cypher_parse_error_context_offset(error);
+
+        asprintf(&errorMsg, "errMsg: %s line: %u, column: %u, offset: %zu errCtx: %s errCtxOffset: %zu", errMsg, errPos.line, errPos.column, errPos.offset, errCtx, errCtxOffset);
```

Comment:
@swilly22 Good catch! Memory leak, and only the last error gets reported.

We have a few options, which I generally dislike:
1) User-level array reply with one error per line. I worry that this may confuse language clients.
2) Concatenate all errors into one string. I don't like this because newlines don't render from the CLI (like the old EXPLAIN returns); it will basically be unreadable.
3) Only report the first error.

I favor option 3 right now, I think. Thoughts?